### ISSUE TITLE
InjectMocks annotation injects fields with non default constructor

### ISSUE THREAD

**Issue by emartynov** [mockito/mockito#714]
Next code started working differently after update to Mockito 2.x.

```
@RunWith(MockitoJUnitRunner.Silent.class)
public class ChangePinActivityPresenterShould
{
  @InjectMocks private ChangePinActivityPresenter presenter;

  @Test
  public void delay_view_finish_until_subscribed()
    throws Exception
  {
    presenter.finishFlow();

    presenter.subscribe(viewMock);

    verify(viewMock).finish();
  }
}

class ChangePinActivityPresenter
  implements ChangePinActivityContract.Presenter
{
  private final Analytics analytics;
  private final PendingActionsProcessor<ChangePinActivityContract.View> pendingActionsProcessor;

  @Nullable private ChangePinActivityContract.View view;
  @Nullable private Action1<ChangePinActivityContract.View> finishAction;

  ChangePinActivityPresenter(
    @NonNull Analytics analytics,
    @NonNull PendingActionsProcessor<ChangePinActivityContract.View> pendingActionsProcessor
  )
  {
    this.analytics = analytics;
    this.pendingActionsProcessor = pendingActionsProcessor;
  }

  @Override
  public void finishFlow()
  {
    if (view != null)
    {
      view.finish();
    }
    else
    {
      finishAction = ChangePinActivityContract.View::finish;
    }
  }

  @Override
  public void subscribe(@NonNull ChangePinActivityContract.View view)
  {
    this.view = view;

    if (finishAction != null)
    {
      finishAction.call(view);
    }
    else
    {
      pendingActionsProcessor.applyAllPendingChanges(view);
    }
  }
}
```

What I observing that **Mockito** uses non-default constructor (or maybe field injection) and field injection (or just it). The test above expects that after test setup `view` field is null. I read [javadoc](http://site.mockito.org/mockito/docs/current/org/mockito/InjectMocks.html) but looks like I have a combination of final/non-final fields and this case is not explained clearly.

I use Java8 on Mack Sierra and my Mockito version is 2.2.0.

**Comment by bric3** [comment]
Hi, 

Mockito injection has been tweaked in version 2. But this should not happen as mockito injects only declared mocks in the test class, but the shown code does not have any. The `view` instance comes from somewhere else. I can't investigate further or reproduce the behavior because the pasted code is incomplete.

**Comment by emartynov** [comment]
Hi guys,

Sorry for the delay in the update.

So I tried to create a project to reproduce the issue two times. But I failed. `@InjectMocks` worked as expected.

I came back to my project and found that it is working as expected. The confusion came because before field injection didn't happen to us. So it was unexpected behaviour for us which is well documented.

So I'm closing the issue. And I will probably create another one to add an ability to disable field injection or control injection in the more explicit way.

Thank you!
